packages/truffle-reporters/Reporter.js

```diff
@@ -1,11 +1,13 @@
 class Reporter {
-  constructor({ options, emitter }) {
+  // const supportedEvents = { compile: compileEvents, unbox: unboxEvents };
```

Comment:
Challenge: can we make it so that we don't have to know the possible events in advance?

Reply:
Did you have any ideas for this? It seems like we'll have to know what events exist in order to match against to set up the handlers, right?

Reply:
So on initialization of a reporter, the event names are split up into two categories: one for names without wildcard characters (`*` or `**`) and another for names that do contain them.
The names without wildcards subscribe directly to whatever event they are named. The names with wildcards have a table built for them in which a regex can be looked up that will match the appropriate events. Whenever any event is fired, the reporter looks through the table of regexes to see if any of the events match the fired event. If it does, it runs the appropriate handler.